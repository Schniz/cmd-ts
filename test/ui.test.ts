import path from "node:path";
import { describe, expect, test } from "vitest";
import { app } from "./util";

test("help for subcommands", async () => {
	const result = await runApp1(["--help"]);
	expect(result.all).toMatchSnapshot();
	expect(result.exitCode).toBe(1);
});

test("invalid subcommand", async () => {
	const result = await runApp1(["subcommand-that-doesnt-exist"]);
	expect(result.all).toMatchSnapshot();
	expect(result.exitCode).toBe(1);
});

test("help for complex command", async () => {
	const result = await runApp1(["complex", "--help"]);
	expect(result.all).toMatchSnapshot();
	expect(result.exitCode).toBe(1);
});

test("too many arguments", async () => {
	const result = await runApp1([
		"--this=will-be-an-error",
		"cat",
		path.relative(process.cwd(), path.join(__dirname, "../package.json")),
		"--and-also-this",
	]);
	expect(result.all).toMatchSnapshot();
	expect(result.exitCode).toBe(1);
});

test("suggests a subcommand on typo", async () => {
	const result = await runApp1(["greek"]);
	expect(result.all).toMatchSnapshot();
	expect(result.exitCode).toBe(1);
});

test("displays subcommand help if no arguments passed", async () => {
	const result = await runApp1([]);
	expect(result.all).toMatchSnapshot();
	expect(result.exitCode).toBe(1);
});

test("displays nested subcommand help if no arguments passed", async () => {
	const result = await runApp1(["composed"]);
	expect(result.all).toMatchSnapshot();
	expect(result.exitCode).toBe(1);
});

test("composes errors", async () => {
	const result = await runApp1([
		"greet",
		"--times=not-a-number",
		"not-capitalized",
	]);
	expect(result.all).toMatchSnapshot();
	expect(result.exitCode).toBe(1);
});

test("multiline error", async () => {
	const result = await runApp1(["greet", "Bon Jovi"]);
	expect(result.all).toMatchSnapshot();
	expect(result.exitCode).toBe(1);
});

test("help for composed subcommands", async () => {
	const result = await runApp1(["composed", "--help"]);
	expect(result.all).toMatchSnapshot();
	expect(result.exitCode).toBe(1);
});

test("help for composed subcommand", async () => {
	const result = await runApp1(["composed", "cat", "--help"]);
	expect(result.all).toMatchSnapshot();
	expect(result.exitCode).toBe(1);
});

test("asynchronous type conversion works for failures", async () => {
	const result = await runApp1(["composed", "cat", "https://httpstat.us/404"]);
	expect(result.all).toMatchSnapshot();
	expect(result.exitCode).toBe(1);
});

test("asynchronous type conversion works for success", async () => {
	const result = await runApp1(["composed", "cat", "https://httpstat.us/200"]);
	expect(result.all).toMatchSnapshot();
	expect(result.exitCode).toBe(0);
});

test("subcommands show their version", async () => {
	const result = await runApp1(["--version"]);
	expect(result.all).toMatchSnapshot();
	expect(result.exitCode).toBe(0);
});

test("failures in defaultValue", async () => {
	const result = await runApp2([]);
	expect(result.all).toMatchSnapshot();
	expect(result.exitCode).toBe(1);
});

test("subcommands with process.argv.slice(2)", async () => {
	const result = await runApp3(["--help"]);
	expect(result.all).toMatchSnapshot();
	expect(result.exitCode).toBe(1);
});

describe("allows positional arguments", () => {
	test("help shows them", async () => {
		const result = await runApp3(["sub2", "--help"]);
		expect(result.all).toMatchSnapshot();
		expect(result.exitCode).toBe(1);
	});

	test("no positionals => all default", async () => {
		const result = await runApp3(["sub2"]);
		expect(result.all).toMatchInlineSnapshot(
			`"{ name: 'anonymous', age: undefined }"`,
		);
		expect(result.exitCode).toBe(0);
	});

	test("expects the correct order", async () => {
		// should fail because we get an age first and `hello` is not a number
		const result = await runApp3(["sub2", "hello"]);
		expect(result.all).toMatchSnapshot();
		expect(result.exitCode).toBe(1);
	});

	test("can take all the arguments", async () => {
		// should fail because we get an age first and `hello` is not a number
		const result = await runApp3(["sub2", "10", "ben"]);
		expect(result.all).toMatchInlineSnapshot(`"{ name: 'ben', age: 10 }"`);
		expect(result.exitCode).toBe(0);
	});
});

const runApp1 = app(path.join(__dirname, "../example/app.ts"));
const runApp2 = app(path.join(__dirname, "../example/app2.ts"));
const runApp3 = app(path.join(__dirname, "../example/app3.ts"));
