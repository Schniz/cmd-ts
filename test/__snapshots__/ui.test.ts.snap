// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`allows positional arguments > expects the correct order 1`] = `
"[31m[1merror: [22m[39mfound [33m1[39m error

  [2msub2[22m [31mhello[39m
[31m       [1m^[22m Not a number[39m


[31m[1mhint: [22m[39mfor more information, try '[33msubcmds sub2 --help[39m'"
`;

exports[`allows positional arguments > help shows them 1`] = `
"[1msubcmds sub2[22m

ARGUMENTS:
  [number] - a number[2m [optional][22m
  [str]    - a string[2m [default: [3manonymous[23m][22m

FLAGS:
  --help, -h - show help"
`;

exports[`asynchronous type conversion works for failures 1`] = `
"[31m[1merror: [22m[39mfound [33m1[39m error

  [2msubcmds[22m [2mcomposed[22m [2mcat[22m [31mhttps://httpstat.us/404[39m
[31m                       [1m^[22m Got status Not Found 404 reading URL[39m


[31m[1mhint: [22m[39mfor more information, try '[33msubcmds composed cat --help[39m'"
`;

exports[`asynchronous type conversion works for success 1`] = `"200 OK"`;

exports[`composes errors 1`] = `
"[31m[1merror: [22m[39mfound [33m3[39m errors

  [2msubcmds[22m [2mgreet[22m [31m--times=not-a-number[39m [2mnot-capitalized[22m
[31m                [1m^[22m Not a number[39m

  [2msubcmds[22m [2mgreet[22m [2m--times=not-a-number[22m [31mnot-capitalized[39m
[31m                                     [1m^[22m name must be capitalized[39m

Along with the following error:
  [31m[1m3.[22m[39m [31mNo value provided for --greeting[39m

[31m[1mhint: [22m[39mfor more information, try '[33msubcmds greet --help[39m'"
`;

exports[`displays nested subcommand help if no arguments passed 1`] = `
"[1msubcmds composed[3m <subcommand>[23m[22m
[2m> [22ma nested subcommand

where [3m<subcommand>[23m can be one of:

[2m- [22mcat - A simple \`cat\` clone [2m[alias: read][22m

[2mFor more help, try running \`[33msubcmds composed <subcommand> --help[39m\`[22m"
`;

exports[`displays subcommand help if no arguments passed 1`] = `
"[1msubcmds[3m <subcommand>[23m[22m
[2m> [22mAn awesome subcommand app!

where [3m<subcommand>[23m can be one of:

[2m- [22mcomplex - Just prints the arguments
[2m- [22mcat - A simple \`cat\` clone [2m[alias: read][22m
[2m- [22mgreet - greet a person
[2m- [22mcomposed - a nested subcommand

[2mFor more help, try running \`[33msubcmds <subcommand> --help[39m\`[22m"
`;

exports[`failures in defaultValue 1`] = `
"[31m[1merror: [22m[39mfound [33m3[39m errors

  [31m[1m1.[22m[39m [31mNo value provided for --user[39m
  [31m[1m2.[22m[39m [31mNo value provided for --password[39m
  [31m[1m3.[22m[39m [31mDefault value not found for '--repo': Can't infer repo from git[39m

[31m[1mhint: [22m[39mfor more information, try '[33mbuild --help[39m'"
`;

exports[`help for complex command 1`] = `
"[1msubcmds complex[22m [2m6.6.6-alpha[22m
[2m> [22mJust prints the arguments

ARGUMENTS:
  <pos1>   - a number
  <str>    - a string
  [...str] - a string

OPTIONS:
  --number, -n <number>         - a number
  --int-or-string <str>         - a string
  --float-or-string <str>       - a string
  --optional-option <str>       - a string[2m [optional][22m
  --no-type-option <str>        - a string
  --optional-with-default <str> - a string[2m [env: SOME_ENV_VAR][22m[2m [default: [3mHello[23m][22m

FLAGS:
  --boolean           - a boolean
  --bool-without-type - a boolean
  --help, -h          - show help
  --version, -v       - print the version"
`;

exports[`help for composed subcommand 1`] = `
"[1msubcmds composed cat[22m
[2m> [22mA simple \`cat\` clone

ARGUMENTS:
  <stream>    - A file path or a URL to make a GET request to
  [...stream] - A file path or a URL to make a GET request to

FLAGS:
  --help, -h - show help"
`;

exports[`help for composed subcommands 1`] = `
"[1msubcmds composed[3m <subcommand>[23m[22m
[2m> [22ma nested subcommand

where [3m<subcommand>[23m can be one of:

[2m- [22mcat - A simple \`cat\` clone [2m[alias: read][22m

[2mFor more help, try running \`[33msubcmds composed <subcommand> --help[39m\`[22m"
`;

exports[`help for subcommands 1`] = `
"[1msubcmds[3m <subcommand>[23m[22m
[2m> [22mAn awesome subcommand app!

where [3m<subcommand>[23m can be one of:

[2m- [22mcomplex - Just prints the arguments
[2m- [22mcat - A simple \`cat\` clone [2m[alias: read][22m
[2m- [22mgreet - greet a person
[2m- [22mcomposed - a nested subcommand

[2mFor more help, try running \`[33msubcmds <subcommand> --help[39m\`[22m"
`;

exports[`invalid subcommand 1`] = `
"[31m[1merror: [22m[39mfound [33m1[39m error

  [2msubcmds[22m [31msubcommand-that-doesnt-exist[39m
[31m          [1m^[22m Not a valid subcommand name[39m


[31m[1mhint: [22m[39mfor more information, try '[33msubcmds --help[39m'"
`;

exports[`multiline error 1`] = `
"[31m[1merror: [22m[39mfound [33m2[39m errors

  [2msubcmds[22m [2mgreet[22m [31mBon Jovi[39m
[31m                [1m^[22m Woah, we're half way there[39m
[31m                  Woah! living on a prayer![39m

Along with the following error:
  [31m[1m2.[22m[39m [31mNo value provided for --greeting[39m

[31m[1mhint: [22m[39mfor more information, try '[33msubcmds greet --help[39m'"
`;

exports[`subcommands show their version 1`] = `"1.0.0"`;

exports[`subcommands with process.argv.slice(2) 1`] = `
"[1msubcmds[3m <subcommand>[23m[22m

where [3m<subcommand>[23m can be one of:

[2m- [22msub1
[2m- [22msub2

[2mFor more help, try running \`[33msubcmds <subcommand> --help[39m\`[22m"
`;

exports[`suggests a subcommand on typo 1`] = `
"[31m[1merror: [22m[39mfound [33m1[39m error

  [2msubcmds[22m [31mgreek[39m
[31m          [1m^[22m Not a valid subcommand name[39m
[31m            Did you mean [3mgreet[23m?[39m


[31m[1mhint: [22m[39mfor more information, try '[33msubcmds --help[39m'"
`;

exports[`too many arguments 1`] = `
"[31m[1merror: [22m[39mfound [33m1[39m error

  [2msubcmds[22m [31m--this=will-be-an-error[39m [2mcat[22m [2mpackage.json[22m [31m--and-also-this[39m
[31m          [1m^[22m Unknown arguments[39m


[31m[1mhint: [22m[39mfor more information, try '[33msubcmds cat --help[39m'"
`;
